source/common/router/req_header_formatter.cc

```diff
@@ -0,0 +1,120 @@
+#include "common/router/req_header_formatter.h"
+
+#include <cstdint>
+#include <iostream>
+#include <string>
+#include <unordered_map>
+#include <vector>
+
+#include "common/config/rds_json.h"
+#include "common/http/access_log/access_log_formatter.h"
+#include "common/http/headers.h"
+#include "common/json/json_loader.h"
+
+namespace Envoy {
+namespace Router {
+
+HeaderFormatterPtr RequestHeaderParser::parseInternal(const std::string& format) {
+  std::string variable_name;
+  if (format.find("%") == 0) {
+    size_t last_occ_pos = format.rfind("%");
+    if (last_occ_pos == std::string::npos || last_occ_pos <= 1) {
+      throw EnvoyException(fmt::format(
+          "Incorrect configuration: {}. Expected the variable to be of format %<variable_name>%",
+          format));
+    }
+    variable_name = format.substr(1, last_occ_pos - 1);
+
+  } else {
+    HeaderFormatterPtr plain_header_formatter_ptr(new PlainHeaderFormatter(format));
+    return plain_header_formatter_ptr;
+  }
+  HeaderFormatterPtr request_header_formatter_ptr(new RequestHeaderFormatter(variable_name));
+  return request_header_formatter_ptr;
+}
+
+RequestHeaderParserPtr RequestHeaderParser::parseRoute(const envoy::api::v2::Route& route) {
+  RequestHeaderParserPtr request_header_parser(new RequestHeaderParser());
+  std::unordered_map<Http::LowerCaseString, HeaderFormatterPtr, Http::LowerCaseStringHasher>
+      header_formatter_map;
+  for (const auto& header_value_option : route.route().request_headers_to_add()) {
+    ENVOY_LOG(debug, "adding key {} to header formatter map", header_value_option.header().key());
+    HeaderFormatterPtr header_formatter =
+        RequestHeaderParser::parseInternal(header_value_option.header().value());
+    request_header_parser->header_formatter_map_.emplace(
+        Http::LowerCaseString(header_value_option.header().key()), std::move(header_formatter));
+  }
+
+  return request_header_parser;
+}
+
+RequestHeaderParserPtr
```

Comment:
i believe request_headers_to_add() should return some sort of iterable, can you check that? (not the HeaderValueOption). i think the problem is in the wrong type you are using.